src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs

```diff
@@ -0,0 +1,563 @@
+                        }
+                        break;
+                    default:
+                        break;
+                }
+            }
+            catch (Exception)
+            {
+                return MsQuicConstants.InternalError;
+            }
+
+            return status;
+        }
+
+        private void HandleEventRecv(ref MsQuicNativeMethods.StreamEvent evt)
+        {
+            static unsafe void CopyToBuffer(Span<byte> buffer, StreamEvent evt)
+            {
+                int length = (int)evt.Data.Recv.Buffers[0].Length;
+                new Span<byte>(evt.Data.Recv.Buffers[0].Buffer, length).CopyTo(buffer);
+            }
+
+            // Need to think hard about backpressure here.
+            // Can this pipe grow infinitely if the consumer isn't reading large chunks?
+            // TODO add a test which does a bunch of large writes and small reads.
+            PipeWriter input = _requestPipe.Writer;
+            int length = (int)evt.Data.Recv.TotalBufferLength;
+            Span<byte> result = input.GetSpan(length);
+            CopyToBuffer(result, evt);
+
+            input.Advance(length);
+
+            ValueTask<FlushResult> flushTask = input.FlushAsync();
+
+            if (!flushTask.IsCompletedSuccessfully)
+            {
+                DisableReceive();
+                ReceiveComplete(0);
+                _ = AwaitFlush(flushTask);
+                return;
+            }
+
+            async Task AwaitFlush(ValueTask<FlushResult> ft)
+            {
+                await ft;
+                EnableReceive();
+                ReceiveComplete(length);
+            }
+        }
+
+        private uint HandleEventPeerRecvAbort()
+        {
+            return MsQuicConstants.Success;
+        }
+
+        private uint HandleEventPeerSendAbort()
+        {
+            return MsQuicConstants.Success;
+        }
+
+        private uint HandleStartComplete()
+        {
+            _started = StartState.Finished;
+            _sendResettableCompletionSource.Complete(MsQuicConstants.Success);
+            return MsQuicConstants.Success;
+        }
+
+        private uint HandleEventSendShutdownComplete(ref MsQuicNativeMethods.StreamEvent evt)
+        {
+            return MsQuicConstants.Success;
+        }
+
+        private uint HandleEventShutdownComplete()
+        {
+            // TODO use another cts here.
+            _sendResettableCompletionSource.Complete(MsQuicConstants.Success);
+            return MsQuicConstants.Success;
+        }
+
+        private uint HandleEventPeerSendClose()
+        {
+            _requestPipe.Writer.Complete();
+            return MsQuicConstants.Success;
+        }
+
+        private uint HandleEventSendComplete(ref MsQuicNativeMethods.StreamEvent evt)
+        {
+            _sendHandle.Free();
+            foreach (MemoryHandle gchBufferArray in _bufferArrays)
+            {
+                gchBufferArray.Dispose();
```

Comment:
This code seems to assume only one outstanding send at a time, since it's cleaning up all the buffers. That doesn't seem like a great design.

Reply:
Why? I don't seem many use cases for starting multiple sends at the same time. That would mean you are doing:
```c#
Task sendTask = stream.SendAsync(buffer);
Task sendTask2 = stream.SendAsync(buffer2);

await sendTask;
await sendTask2;
```
Pretty much in all cases, calling SendAsync multiple times in a row without awaiting isn't guaranteed to work.

Reply:
FWIW `Socket` does allow multiple outstanding sends.

I would wouldn't push for the same functionality here, though.

Reply:
QUIC allows for as many parallel sends on a single stream as you'd like.

Reply:
Makes sense. I'll leave it as a TODO to consider allowing concurrent sends.